abi.go

```
@@ -0,0 +1,446 @@
+		a.Events = map[string]*Event{}
+	}
+	name := overloadedName(e.Name, func(s string) bool {
+		_, ok := a.Events[s]
+		return ok
+	})
+	a.Events[name] = e
+}
+
+func (a *ABI) addMethod(m *Method) {
+	if len(a.Methods) == 0 {
+		a.Methods = map[string]*Method{}
+	}
+	if len(a.MethodsBySignature) == 0 {
+		a.MethodsBySignature = map[string]*Method{}
+	}
+	name := overloadedName(m.Name, func(s string) bool {
+		_, ok := a.Methods[s]
+		return ok
+	})
+	a.Methods[name] = m
+	a.MethodsBySignature[m.Sig()] = m
+}
+
+func overloadedName(rawName string, isAvail func(string) bool) string {
+	name := rawName
+	ok := isAvail(name)
+	for idx := 0; ok; idx++ {
+		name = fmt.Sprintf("%s%d", rawName, idx)
+		ok = isAvail(name)
+	}
+	return name
+}
+
+// NewABI returns a parsed ABI struct
+func NewABI(s string) (*ABI, error) {
+	return NewABIFromReader(bytes.NewReader([]byte(s)))
+}
+
+// MustNewABI returns a parsed ABI contract or panics if fails
+func MustNewABI(s string) *ABI {
+	a, err := NewABI(s)
+	if err != nil {
+		panic(err)
+	}
+	return a
+}
+
+// NewABIFromReader returns an ABI object from a reader
+func NewABIFromReader(r io.Reader) (*ABI, error) {
+	var abi *ABI
+	dec := json.NewDecoder(r)
+	if err := dec.Decode(&abi); err != nil {
+		return nil, err
+	}
+	return abi, nil
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface
+// nolint
+func (a *ABI) UnmarshalJSON(data []byte) error {
+	var fields []struct {
+		Type            string
+		Name            string
+		Constant        bool
+		Anonymous       bool
+		StateMutability string
+		Inputs          []*ArgumentStr
+		Outputs         []*ArgumentStr
+	}
+
+	if err := json.Unmarshal(data, &fields); err != nil {
+		return err
+	}
+
+	for _, field := range fields {
+		switch field.Type {
+		case "constructor":
+			if a.Constructor != nil {
+				return fmt.Errorf("multiple constructor declaration")
+			}
+			input, err := NewTupleTypeFromArgs(field.Inputs)
+			if err != nil {
+				panic(err)
```

Should we panic here? Why not just propagate an error?

---

fixed.